tests/test.py

```diff
@@ -9859,6 +9861,311 @@ def testTransactionRollback(self):
+                SELECT value_bytes
+                FROM test_proper_db_flushing_on_error
+                WHERE id = $1 OR id = $2
+            """
+
+            stmt = await conn.prepare(select_two_query)
+
+            with self.assertRaises(asyncpg.exceptions.SyntaxOrAccessError) as ex:
+                await stmt.fetch(corrupted_data['id'], data['id'])
+
+            self.assertEqual(ex.exception.message,
+                             'encoding error in column "value_bytes"')
+
+            # Insert and select new data using the same connection to be sure
+            # it doesn't close or get out of sync
+            data = {
+                'id': get_random_id(),
+                'value_bytes': random_bytes(),
+            }
+            await conn.execute(insert_query, data['id'], data['value_bytes'])
+            row = await conn.fetchrow(select_query, data['id'])
+            self.assertEqual(data['value_bytes'], row['value_bytes'])
+
+        loop = asyncio.get_event_loop()
+        loop.run_until_complete(test())
+
+    def testTransactionPreparedRollback(self):
+        """
+        Test that connection is not closed in case of "encoding error" with
+        prepared statement and a driver can do rollback in transaction after
+        that.
+        """
+        async def test():
+            self.encryptor_table.create(bind=self.engine_raw, checkfirst=True)
+            # Insert data that will trigger decryption error
+            corrupted_data = {
+                'id': get_random_id(),
+                'value_bytes': random_bytes(),
+            }
+            self.engine_raw.execute(
+                self.encryptor_table.insert(), corrupted_data)
+            data = {
+                'id': get_random_id(),
+                'value_bytes': random_bytes(),
+            }
+
+            conn = await self.executor.connect()
+
+            insert_query = """
+                INSERT INTO test_proper_db_flushing_on_error(id, value_bytes)
+                VALUES ($1, $2)
+            """
+            select_query = """
+                SELECT value_bytes
+                FROM test_proper_db_flushing_on_error
+                WHERE id = $1
+            """
+
+            with self.assertRaises(asyncpg.exceptions.SyntaxOrAccessError) as ex:
+                async with conn.transaction():
+                    await conn.execute(insert_query, data['id'], data['value_bytes'])
+
+                    row = await conn.fetchrow(select_query, data['id'])
+                    self.assertEqual(data['value_bytes'], row['value_bytes'])
+
+                    stmt = await conn.prepare(select_query)
+                    # Expect encoding error
+                    await stmt.fetch(corrupted_data['id'])
+
+            self.assertEqual(ex.exception.message,
+                             'encoding error in column "value_bytes"')
+            row = await conn.fetchrow(select_query, data['id'])
+            self.assertEqual(row, None)
```

Comment:
hm, can you explain why we expect `None` here? As I understand:
1. we inserted data and it is already in the database
2. We tried to get this data in the transaction, took error from acra side and handled exception
3. We try to get same data but get None as result, not exception due to same error and not error that `select` returns no data?

Reply:
oh, thanks. can you add some comment above this check with explanation why we expect here `None`?
Because I'm sure that someone will forget that most of ORM use `ROLLBACK` on any exception. Or mention that after that we expect that ORM will do ROLLBACK. Just for case if event this ORM will change his logic and stop do ROLLBACK on errors.